File: src/ios/Agent/Backup/BackupCrypto.swift
import CommonCrypto
import CryptoKit
import Foundation

private let logger = AppLogger(category: "Backup")

/// Passphrase-based encryption for `.minisbak` packages
/// (docs/backup-restore-design.md §5, scheme `minisbak-enc/1`).
///
/// ## Chosen parameters, and the Android equivalent for each
///
/// Selection rule (§5.2's "全部算法双端标准库/成熟库可得", reaffirmed by the
/// 2026-08-14 cross-platform constraint): **an algorithm both platforms can do
/// with a standard library beats one that is better on a single platform.**
/// Nothing below needs a third-party dependency on either side.
///
/// | Parameter | iOS (this file) | Android equivalent |
/// |---|---|---|
/// | KDF | `CCKeyDerivationPBKDF(kCCPBKDF2, kCCPRFHmacAlgSHA256)` — CommonCrypto | `SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")` + `PBEKeySpec(pw, salt, 600_000, 256)` — JCA, API 26+ |
/// | Iterations | 600 000 | identical — it is part of the wire format (`kdf.iterations`) |
/// | Salt | 16 random bytes, `SecRandomCopyBytes` | `SecureRandom().nextBytes(ByteArray(16))` |
/// | Subkey derivation | `HKDF<SHA256>.deriveKey` — CryptoKit | `Hkdf` from Tink, or ~15 lines of `javax.crypto.Mac("HmacSHA256")` |
/// | Cipher | `AES.GCM.seal/open` — CryptoKit | `Cipher.getInstance("AES/GCM/NoPadding")` + `GCMParameterSpec(128, nonce)` — JCA |
/// | Nonce | 12 bytes, CryptoKit-generated per segment | same; GCM standard, `SecureRandom` |
/// | Tag | 128-bit, appended by `sealed.combined` | `GCMParameterSpec(128, …)` — JCA appends it identically |
/// | Segment | 4 MiB, `UInt32` big-endian length prefix | `ByteBuffer.order(ByteOrder.BIG_ENDIAN)` |
/// | AAD | UTF-8 `"<path>#<segment>"` | `cipher.updateAAD(s.toByteArray(UTF_8))` |
///
/// **Why not Argon2id**, which §5.2 lists first: no Argon2 ships in the Apple
/// SDKs, and Android has no platform Argon2 either — both sides would need a
/// third-party dependency (argon2-jvm / libsodium-jni on Android, a vendored C
/// library on iOS, in an app whose native-dependency story is already fragile —
/// its libish_emu.a is device-only and cannot link for Simulator at all). Two
/// new native dependencies to align, versus zero: PBKDF2 is the lower-risk
/// choice, and §5.2 explicitly sanctions it. 600k iterations is OWASP's current
/// PBKDF2-SHA256 floor.
///
/// The format carries `kdf.alg`, so **both KDFs stay decodable by
/// construction** — if Argon2 is adopted later, new packages declare
/// `argon2id` and old ones keep opening. An unknown `alg` is refused loudly
/// rather than guessed at.
enum BackupCrypto {

    static let scheme = "minisbak-enc/1"
    static let pbkdf2Iterations = 600_000
    static let saltBytes = 16

    /// Payload is split into independently sealed 4MiB segments (§5.3) so a
    /// multi-hundred-MB blob never has to be held in memory to encrypt or
    /// decrypt, and so truncating or reordering segments is detectable.
    static let segmentSize = 4 * 1024 * 1024

    /// `MBK1` — lets a reader identify an encrypted member before trying to
    /// parse it as JSON.
    static let magic = Data([0x4D, 0x42, 0x4B, 0x31])

    // MARK: - Key hierarchy (§5.2)

    /// The derived key set. Lives only for the duration of an export/import and
    /// is zeroed on deinit; the passphrase itself is never persisted.
    final class Keys {
        /// `data/*` and `blobs/*`.
        let dataKey: SymmetricKey
        /// `secrets.json` only — the separate subkey is what makes "strip the
        /// credentials from this package" a file removal rather than a re-encrypt.
        let secretsKey: SymmetricKey
        /// Authenticates manifest.json.
        let macKey: SymmetricKey
        /// Answers "is this passphrase right?" without touching any payload.
        let verifierKey: SymmetricKey

        fileprivate init(kek: SymmetricKey) {
            func sub(_ info: String) -> SymmetricKey {
                SymmetricKey(data: HKDF<SHA256>.deriveKey(
                    inputKeyMaterial: kek,
                    info: Data(info.utf8),
                    outputByteCount: 32))
            }
            dataKey = sub("minisbak/data")
            secretsKey = sub("minisbak/secrets")
            macKey = sub("minisbak/mac")
            verifierKey = sub("minisbak/verify")
        }

        /// `HMAC(verifier_key, "minisbak-v1")` truncated to 16 bytes (§5.2).
        /// Written to the manifest at export; recomputed at import so a wrong
        /// passphrase fails instantly instead of surfacing as a decrypt error
        /// halfway through a multi-GB restore.
        var verifier: String {
            var mac = HMAC<SHA256>(key: verifierKey)
            mac.update(data: Data("minisbak-v1".utf8))
            return Data(mac.finalize().prefix(16)).base64EncodedString()
        }
    }

    enum CryptoError: LocalizedError {
        case unsupportedKDF(String)
        case unsupportedScheme(String)
        case wrongPassphrase
        case corruptMember(String)
        case manifestTampered

        var errorDescription: String? {
            switch self {
            case .unsupportedKDF(let a):
                return "This backup uses an unsupported key-derivation algorithm (\(a)). Please update the app."
            case .unsupportedScheme(let s):
                return "This backup uses an unsupported encryption scheme (\(s)). Please update the app."
            case .wrongPassphrase:
                return "Incorrect passphrase."
            case .corruptMember(let p):
                return "Encrypted content is damaged or was tampered with: \(p)"
            case .manifestTampered:
                return "The backup's manifest failed authentication — it may have been modified."
            }
        }
    }

    // MARK: - Derivation

    static func makeSalt() -> Data {
        var bytes = [UInt8](repeating: 0, count: saltBytes)
        _ = SecRandomCopyBytes(kSecRandomDefault, saltBytes, &bytes)
        return Data(bytes)
    }

    /// Derive the key set from a passphrase. Deliberately synchronous and slow —
    /// 600k PBKDF2 rounds take roughly a second on an A13, which is the point.
    static func deriveKeys(passphrase: String, kdf: BackupManifest.Encryption.KDF) throws -> Keys {
        guard let salt = Data(base64Encoded: kdf.salt) else {
            throw CryptoError.corruptMember("manifest.encryption.kdf.salt")
        }
        switch kdf.alg {
        case "pbkdf2-hmac-sha256":
            let iterations = kdf.iterations ?? pbkdf2Iterations
            return Keys(kek: try pbkdf2(passphrase: passphrase, salt: salt, iterations: iterations))
        case "argon2id":
            // Written by a future build that vendors Argon2. Refusing loudly is
            // correct: a silent fallback to PBKDF2 would derive a different key
            // and surface as "wrong passphrase", sending the user hunting for a
            // password problem that doesn't exist.
            throw CryptoError.unsupportedKDF(kdf.alg)
        default:
            throw CryptoError.unsupportedKDF(kdf.alg)
        }
    }

    /// The KDF descriptor to write into a new package.
    static func currentKDF(salt: Data) -> BackupManifest.Encryption.KDF {
        BackupManifest.Encryption.KDF(
            alg: "pbkdf2-hmac-sha256", mKib: nil, t: nil, p: nil,
            iterations: pbkdf2Iterations, salt: salt.base64EncodedString())
    }

    private static func pbkdf2(passphrase: String, salt: Data, iterations: Int) throws -> SymmetricKey {
        var out = [UInt8](repeating: 0, count: 32)
        let pwBytes = Array(passphrase.utf8)
        let status: Int32 = salt.withUnsafeBytes { saltPtr in
            pwBytes.withUnsafeBufferPointer { pwPtr in
                CCKeyDerivationPBKDF(
                    CCPBKDFAlgorithm(kCCPBKDF2),
                    // The passphrase is passed as bytes with an explicit length,
                    // so an embedded NUL can't truncate it the way a C-string
                    // API would.
                    UnsafeRawPointer(pwPtr.baseAddress!).assumingMemoryBound(to: CChar.self),
                    pwBytes.count,
                    saltPtr.bindMemory(to: UInt8.self).baseAddress!, salt.count,
                    CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                    UInt32(iterations),
                    &out, out.count)
            }
        }
        guard status == kCCSuccess else {
            throw CryptoError.corruptMember("key derivation failed (status \(status))")
        }
        return SymmetricKey(data: Data(out))
    }

    // MARK: - Member encryption (§5.3)

    /// Encrypt one package member, streaming through 4MiB segments.
    ///
    /// Layout: `MBK1` then, per segment, `UInt32 big-endian length` followed by
    /// the AES-GCM sealed box (nonce ‖ ciphertext ‖ tag).
    ///
    /// The AAD binds each segment to **both** its path and its index, so a
    /// member cannot be renamed to impersonate another (swapping an old
    /// `sessions.jsonl.enc` in as `skills.jsonl.enc`) and segments cannot be
    /// dropped or reordered within a member.
    static func encryptFile(at source: URL, to destination: URL,
                            key: SymmetricKey, path: String) throws {
        let input = try FileHandle(forReadingFrom: source)
        defer { try? input.close() }
        FileManager.default.createFile(atPath: destination.path, contents: nil)
        let output = try FileHandle(forWritingTo: destination)
        defer { try? output.close() }

        try output.write(contentsOf: magic)
        var index = 0
        while true {
            // [T-backup-scan-jetsam] Per-segment pool: FileHandle.read hands
            // back autoreleased NSData, and seal() allocates a same-sized
            // ciphertext — without draining per segment, encrypting a large
            // member holds every segment of it in memory at once.
            let done = try autoreleasepool { () -> Bool in
                let chunk = try input.read(upToCount: segmentSize) ?? Data()
                // A zero-length file still writes zero segments; the magic alone
                // marks it as encrypted, and decryption yields an empty file.
                if chunk.isEmpty { return true }
                let sealed = try AES.GCM.seal(chunk, using: key,
                                              authenticating: aad(path: path, segment: index))
                guard let combined = sealed.combined else {
                    throw CryptoError.corruptMember(path)
                }
                var length = UInt32(combined.count).bigEndian
                try output.write(contentsOf: Data(bytes: &length, count: 4))
                try output.write(contentsOf: combined)
                index += 1
                return false
            }
            if done { break }
        }
    }

    static func decryptFile(at source: URL, to destination: URL,
                            key: SymmetricKey, path: String) throws {
        let input = try FileHandle(forReadingFrom: source)
        defer { try? input.close() }
        guard let header = try input.read(upToCount: magic.count), header == magic else {
            throw CryptoError.corruptMember(path)
        }
        FileManager.default.createFile(atPath: destination.path, contents: nil)
        let output = try FileHandle(forWritingTo: destination)
        defer { try? output.close() }

        var index = 0
        while true {
            // [T-backup-scan-jetsam] Same per-segment pool as encryptFile.
            let done = try autoreleasepool { () -> Bool in
                guard let lenData = try input.read(upToCount: 4), !lenData.isEmpty else { return true }
                guard lenData.count == 4 else { throw CryptoError.corruptMember(path) }
                let length = Int(lenData.withUnsafeBytes { $0.loadUnaligned(as: UInt32.self).bigEndian })
                guard let body = try input.read(upToCount: length), body.count == length else {
                    throw CryptoError.corruptMember(path)
                }
                do {
                    let box = try AES.GCM.SealedBox(combined: body)
                    let plain = try AES.GCM.open(box, using: key,
                                                 authenticating: aad(path: path, segment: index))
                    try output.write(contentsOf: plain)
                } catch {
                    // An authentication failure here is indistinguishable from a
                    // wrong key, but the caller checks the verifier first, so by
                    // this point the passphrase is known-good and this really is
                    // damage or tampering.
                    throw CryptoError.corruptMember(path)
                }
                index += 1
                return false
            }
            if done { break }
        }
    }

    private static func aad(path: String, segment: Int) -> Data {
        Data("\(path)#\(segment)".utf8)
    }

    // MARK: - Manifest authentication (§5.3)

    /// `HMAC-SHA256(K_mac, canonical_json(manifest without manifest_mac))`.
    ///
    /// Stops an attacker editing the plaintext manifest — e.g. deleting the
    /// `providers` entry to hide that credentials are present, or downgrading
    /// the KDF parameters.
    static func manifestMAC(_ manifest: BackupManifest, key: SymmetricKey) throws -> String {
        var copy = manifest
        copy.manifestMac = nil
        let encoder = JSONEncoder()
        // Sorted keys make the byte sequence canonical, so both ends compute
        // the MAC over identical input.
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        encoder.dateEncodingStrategy = .iso8601
        let data = try encoder.encode(copy)
        var mac = HMAC<SHA256>(key: key)
        mac.update(data: data)
        return Data(mac.finalize()).base64EncodedString()
    }

    /// MAC over the RAW BYTES of manifest.json exactly as written to the
    /// package (stored in the sidecar member `manifest.mac`).
    ///
    /// The struct-based MAC above has a forward-compatibility flaw: it MACs a
    /// re-encoding of the DECODED manifest, so a reader that predates any
    /// later manifest field (e.g. `snapshot_at`) silently drops it on decode,
    /// re-encodes different bytes, and reports "tampered" for a perfectly good
    /// package. Raw bytes never go through a decoder, so unknown fields cost
    /// nothing — this variant stays valid across versions in BOTH directions.
    static func manifestMAC(rawBytes: Data, key: SymmetricKey) -> String {
        var mac = HMAC<SHA256>(key: key)
        mac.update(data: rawBytes)
        return Data(mac.finalize()).base64EncodedString()
    }

    static func verifyManifestMAC(rawBytes: Data, expected: String, key: SymmetricKey) throws {
        let actual = manifestMAC(rawBytes: rawBytes, key: key)
        guard let a = Data(base64Encoded: actual), let b = Data(base64Encoded: expected),
              a.count == b.count,
              a.withUnsafeBytes({ ap in b.withUnsafeBytes { bp in
                  timingsafe_bcmp(ap.baseAddress!, bp.baseAddress!, a.count) == 0
              } }) else {
            throw CryptoError.manifestTampered
        }
    }

    static func verifyManifestMAC(_ manifest: BackupManifest, key: SymmetricKey) throws {
        guard let expected = manifest.manifestMac else { return }
        let actual = try manifestMAC(manifest, key: key)
        // Constant-time compare via CryptoKit's own equality on the raw bytes.
        guard let a = Data(base64Encoded: actual), let b = Data(base64Encoded: expected),
              a.count == b.count,
              a.withUnsafeBytes({ ap in b.withUnsafeBytes { bp in
                  timingsafe_bcmp(ap.baseAddress!, bp.baseAddress!, a.count) == 0
              } }) else {
            throw CryptoError.manifestTampered
        }
    }

    /// Constant-time verifier comparison, so a wrong passphrase can't be
    /// narrowed down by timing.
    static func verifierMatches(_ stored: String, keys: Keys) -> Bool {
        guard let a = Data(base64Encoded: stored),
              let b = Data(base64Encoded: keys.verifier),
              a.count == b.count else { return false }
        return a.withUnsafeBytes { ap in
            b.withUnsafeBytes { bp in
                timingsafe_bcmp(ap.baseAddress!, bp.baseAddress!, a.count) == 0
            }
        }
    }
}

File: src/android/app/src/main/java/com/openminis/app/backup/BackupCrypto.kt
package com.openminis.app.backup

import java.io.File
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
// java.util.Base64 rather than android.util.Base64 (minSdk 26 has it): this
// layer is pure wire format with no framework dependency, so it stays testable
// under plain JVM unit tests — the module sets `unitTests.isReturnDefaultValues`,
// which would make android.util.Base64 silently return null and turn every
// cross-platform vector assertion into a false pass.
import java.util.Base64

/**
 * Passphrase-based encryption for `.minisbak` packages
 * (docs/backup-restore-design.md §5, scheme `minisbak-enc/1`).
 *
 * This is the Android column of the §5.2.1 parameter table, implemented
 * against `src/ios/Agent/Backup/BackupCrypto.swift`. Every value here is wire
 * format — a deviation makes packages mutually unreadable:
 *
 *   - KDF: PBKDF2-HMAC-SHA256, 600 000 iterations, 16-byte salt, 256-bit KEK.
 *   - Subkeys: HKDF-SHA256 (empty salt — identical to CryptoKit's no-salt
 *     form, since HMAC pads both an empty key and 32 zero bytes to the same
 *     block) with infos `minisbak/{data,secrets,mac,verify}`.
 *   - Cipher: AES-256-GCM, 12-byte random nonce, 128-bit tag, wire layout per
 *     segment = `UInt32 BE length ‖ nonce ‖ ciphertext ‖ tag` (the length
 *     covers nonce+ct+tag, matching CryptoKit's `sealed.combined`).
 *   - Segments: 4 MiB plaintext each, AAD = UTF-8 `"<path>#<segment>"` —
 *     binds each segment to its member path AND index, so members can't be
 *     renamed to impersonate each other and segments can't be reordered.
 *   - Member magic: ASCII `MBK1`.
 *
 * Why PBKDF2 and not Argon2id: neither platform ships Argon2 in its standard
 * library, and §5.2.1 rules that dual-platform standard-library availability
 * beats single-platform optimality. `kdf.alg` keeps both decodable by
 * construction; an unknown alg is refused loudly, never silently substituted
 * (a silent fallback derives a different key and masquerades as "wrong
 * passphrase").
 */
object BackupCrypto {

    const val SCHEME = "minisbak-enc/1"
    const val PBKDF2_ITERATIONS = 600_000
    const val SALT_BYTES = 16

    /** 4 MiB plaintext per independently sealed segment (§5.3). */
    const val SEGMENT_SIZE = 4 * 1024 * 1024

    /** `MBK1` — identifies an encrypted member before any JSON parse attempt. */
    val MAGIC = byteArrayOf(0x4D, 0x42, 0x4B, 0x31)

    private const val NONCE_BYTES = 12
    private const val TAG_BITS = 128

    class CryptoException(message: String) : Exception(message)
    class WrongPassphraseException : Exception("Incorrect passphrase.")
    class ManifestTamperedException :
        Exception("The backup's manifest failed authentication — it may have been modified.")
    class CorruptMemberException(path: String) :
        Exception("Encrypted content is damaged or was tampered with: $path")
    class UnsupportedKDFException(alg: String) :
        Exception("This backup uses an unsupported key-derivation algorithm ($alg). Please update the app.")

    // MARK: - Key hierarchy (§5.2)

    /**
     * The derived key set. Lives only for the duration of an export/import;
     * call [destroy] when finished — the passphrase itself is never persisted.
     */
    class Keys internal constructor(kek: ByteArray) {
        /** Everything under `data/` and `blobs/`. */
        val dataKey: ByteArray = hkdfSha256(kek, "minisbak/data")
        /**
         * `secrets.json` only — the separate subkey is what makes "strip the
         * credentials from this package" a file removal, not a re-encrypt.
         */
        val secretsKey: ByteArray = hkdfSha256(kek, "minisbak/secrets")
        /** Authenticates manifest.json. */
        val macKey: ByteArray = hkdfSha256(kek, "minisbak/mac")
        /** Answers "is this passphrase right?" without touching any payload. */
        val verifierKey: ByteArray = hkdfSha256(kek, "minisbak/verify")

        init {
            kek.fill(0)
        }

        /**
         * `HMAC(verifier_key, "minisbak-v1")` truncated to 16 bytes, base64
         * (§5.2). Written to the manifest at export; recomputed at import so a
         * wrong passphrase fails instantly instead of surfacing as a decrypt
         * error halfway through a multi-GB restore.
         */
        val verifier: String
            get() {
                val mac = hmacSha256(verifierKey, "minisbak-v1".toByteArray(Charsets.UTF_8))
                return Base64.getEncoder().encodeToString(mac.copyOf(16))
            }

        fun destroy() {
            dataKey.fill(0); secretsKey.fill(0); macKey.fill(0); verifierKey.fill(0)
        }
    }

    // MARK: - Derivation

    fun makeSalt(): ByteArray = ByteArray(SALT_BYTES).also { SecureRandom().nextBytes(it) }

    /**
     * Derive the key set from a passphrase. Deliberately synchronous and slow —
     * 600k PBKDF2 rounds are the point. Call off the main thread.
     */
    fun deriveKeys(passphrase: String, kdf: BackupManifest.Encryption.KDF): Keys {
        val salt = try {
            Base64.getDecoder().decode(kdf.salt)
        } catch (e: IllegalArgumentException) {
            throw CorruptMemberException("manifest.encryption.kdf.salt")
        }
        return when (kdf.alg) {
            "pbkdf2-hmac-sha256" -> {
                val iterations = kdf.iterations ?: PBKDF2_ITERATIONS
                Keys(pbkdf2(passphrase, salt, iterations))
            }
            // "argon2id" would come from a future build that vendors Argon2;
            // refusing loudly is correct — see the header comment.
            else -> throw UnsupportedKDFException(kdf.alg)
        }
    }

    /** The KDF descriptor to write into a new package. */
    fun currentKDF(salt: ByteArray) = BackupManifest.Encryption.KDF(
        alg = "pbkdf2-hmac-sha256",
        salt = Base64.getEncoder().encodeToString(salt),
        iterations = PBKDF2_ITERATIONS,
    )

    private fun pbkdf2(passphrase: String, salt: ByteArray, iterations: Int): ByteArray {
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val spec = PBEKeySpec(passphrase.toCharArray(), salt, iterations, 256)
        try {
            return factory.generateSecret(spec).encoded
        } finally {
            spec.clearPassword()
        }
    }

    /**
     * HKDF-SHA256, RFC 5869, empty salt, single-block expand (32 bytes out).
     * Matches CryptoKit's `HKDF<SHA256>.deriveKey(inputKeyMaterial:info:)`.
     */
    private fun hkdfSha256(ikm: ByteArray, info: String): ByteArray {
        val prk = hmacSha256(ByteArray(0), ikm) // extract, salt = empty
        return hmacSha256(prk, info.toByteArray(Charsets.UTF_8) + byteArrayOf(0x01))
    }

    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        // HMAC forbids a zero-length key in JCA; pad to the block-size
        // equivalent, which HMAC's own key padding makes identical.
        mac.init(SecretKeySpec(if (key.isEmpty()) ByteArray(32) else key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    // MARK: - Member encryption (§5.3)

    /**
     * Encrypt one package member, streaming through 4 MiB segments.
     *
     * Layout: `MBK1` then, per segment, `UInt32 big-endian length` followed by
     * `nonce ‖ ciphertext ‖ tag`. A zero-length file writes zero segments; the
     * magic alone marks it as encrypted.
     */
    fun encryptFile(source: File, destination: File, key: ByteArray, path: String) {
        source.inputStream().buffered().use { input ->
            destination.outputStream().buffered().use { output ->
                output.write(MAGIC)
                var index = 0
                while (true) {
                    val chunk = readUpTo(input, SEGMENT_SIZE) ?: break
                    val nonce = ByteArray(NONCE_BYTES).also { SecureRandom().nextBytes(it) }
                    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                    cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"),
                        GCMParameterSpec(TAG_BITS, nonce))
                    cipher.updateAAD(aad(path, index))
                    val sealed = cipher.doFinal(chunk)
                    val combined = nonce + sealed
                    output.write(ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN)
                        .putInt(combined.size).array())
                    output.write(combined)
                    index += 1
                }
            }
        }
    }

    fun decryptFile(source: File, destination: File, key: ByteArray, path: String) {
        source.inputStream().buffered().use { input ->
            val header = ByteArray(MAGIC.size)
            if (!readFully(input, header) || !header.contentEquals(MAGIC)) {
                throw CorruptMemberException(path)
            }
            destination.outputStream().buffered().use { output ->
                decryptStream(input, output, key, path)
            }
        }
    }

    /** Decrypt a member whose `MBK1` magic has already been consumed. */
    fun decryptStream(input: InputStream, output: OutputStream, key: ByteArray, path: String) {
        var index = 0
        while (true) {
            val lenBytes = ByteArray(4)
            val first = input.read()
            if (first < 0) break // clean EOF at a segment boundary
            lenBytes[0] = first.toByte()
            if (!readFully(input, lenBytes, offset = 1)) throw CorruptMemberException(path)
            val length = ByteBuffer.wrap(lenBytes).order(ByteOrder.BIG_ENDIAN).int
            if (length < NONCE_BYTES + TAG_BITS / 8 || length > SEGMENT_SIZE + NONCE_BYTES + TAG_BITS / 8 + 64) {
                throw CorruptMemberException(path)
            }
            val body = ByteArray(length)
            if (!readFully(input, body)) throw CorruptMemberException(path)
            try {
                val cipher = Cipher.getInstance("AES/GCM/NoPadding")
                cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"),
                    GCMParameterSpec(TAG_BITS, body, 0, NONCE_BYTES))
                cipher.updateAAD(aad(path, index))
                output.write(cipher.doFinal(body, NONCE_BYTES, body.size - NONCE_BYTES))
            } catch (e: Exception) {
                // An auth failure here is indistinguishable from a wrong key,
                // but the caller checks the verifier first, so by this point
                // the passphrase is known-good and this really is damage.
                throw CorruptMemberException(path)
            }
            index += 1
        }
    }

    private fun aad(path: String, segment: Int): ByteArray =
        "$path#$segment".toByteArray(Charsets.UTF_8)

    // MARK: - Manifest authentication (§5.3)

    /**
     * MAC over the RAW BYTES of manifest.json exactly as written to the
     * package — stored in the sidecar member `manifest.mac`, which both
     * platforms' readers prefer. (The embedded `manifest_mac` field is a
     * Swift-re-encoding MAC that Android neither writes nor can verify;
     * raw bytes never go through a decoder, so unknown fields cost nothing
     * and the sidecar stays valid across versions in both directions.)
     */
    fun manifestMac(rawBytes: ByteArray, key: ByteArray): String =
        Base64.getEncoder().encodeToString(hmacSha256(key, rawBytes))

    fun verifyManifestMac(rawBytes: ByteArray, expected: String, key: ByteArray) {
        val actual = Base64.getDecoder().decode(manifestMac(rawBytes, key))
        val exp = try {
            Base64.getDecoder().decode(expected.trim())
        } catch (e: IllegalArgumentException) {
            throw ManifestTamperedException()
        }
        // MessageDigest.isEqual is constant-time since API 16.
        if (!MessageDigest.isEqual(actual, exp)) throw ManifestTamperedException()
    }

    /** Constant-time verifier comparison, so a wrong passphrase can't be timed. */
    fun verifierMatches(stored: String, keys: Keys): Boolean {
        val a = try {
            Base64.getDecoder().decode(stored)
        } catch (e: IllegalArgumentException) {
            return false
        }
        val b = Base64.getDecoder().decode(keys.verifier)
        return MessageDigest.isEqual(a, b)
    }

    // MARK: - Stream helpers

    /** Read up to [max] bytes; null at clean EOF, never a zero-length array. */
    private fun readUpTo(input: InputStream, max: Int): ByteArray? {
        val buf = ByteArray(max)
        var read = 0
        while (read < max) {
            val n = input.read(buf, read, max - read)
            if (n < 0) break
            read += n
        }
        if (read == 0) return null
        return if (read == max) buf else buf.copyOf(read)
    }

    private fun readFully(input: InputStream, buf: ByteArray, offset: Int = 0): Boolean {
        var read = offset
        while (read < buf.size) {
            val n = input.read(buf, read, buf.size - read)
            if (n < 0) return false
            read += n
        }
        return true
    }
}
